File: lm-js/18-445两数相加II.js
// 445. 两数相加 II
// 给你两个 非空 链表来代表两个非负整数。数字最高位位于链表开始位置。它们的每个节点只存储一位数字。将这两数相加会返回一个新的链表。

// 你可以假设除了数字 0 之外，这两个数字都不会以零开头。

 

// 进阶：

// 如果输入链表不能修改该如何处理？换句话说，你不能对列表中的节点进行翻转。

 

// 示例：

// 输入：(7 -> 2 -> 4 -> 3) + (5 -> 6 -> 4)
// 输出：7 -> 8 -> 0 -> 7



/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    let len=0,newL1=l1,newL2=l2,sum=[]
    //遍历判断链表谁的比较长，长度差为多少
    while(l1){
        l1=l1.next
        len++
    }
    while(l2){
        l2=l2.next
        len--
    }
    while(newL1||newL2){
        if(len>0){//如果l1比较长，就把长度差外面的这些数字，直接push进入和的数组
            while(len)
            {sum.push(newL1.val)
            newL1&&(newL1=newL1.next)
            len--
            }
        }
        else if(len<0){//反之将l2的长度差之外前面的这些数字，push进入和的数组。
            len=Math.abs(len)
            while(len)
            {
                sum.push(newL2.val)
            newL2&&(newL2=newL2.next)
            len--
            }
        }
        else{//剩下就是长度一致，才开始求和，此时不考虑进位，只求和
            (newL1&&newL2)&&sum.push(newL1.val+newL2.val)
            newL1=newL1.next
            newL2=newL2.next
        }
    }
    //利用到了数组存值，然后翻转一下
  sum.reverse()
  let carry=0
  for(let i=0;i<sum.length;i++){//此时再把和超过9的元素，carry进位1，等待下一次遍历再加上carry
      let curSum=sum[i]+carry  //当前和为元素+carry
      sum[i]=curSum>9?curSum%10:curSum //元素对10取余
      carry=curSum>9?1:0 //下一位进位为1
  }
  //这里则是当最高位相加和也进一了，但是已经循环结束了，则新增一个最高位，且值为1
  if(carry==1){
      sum.push(1)
  }
  //当和数组全部调整完毕，再根据pop()来遍历构建一个新的链表
  let res=new ListNode(),newRes=res
  while(sum.length){
      res.next=new ListNode(sum.pop())
      res=res.next
  }
  return newRes.next
};

/**题解
   首先对于长的链表，在长度差之外这些数字即为求和之后的值(即len之前，和即本身)，然后遍历两个数组，进行求和赋值
   此题的关键是我借用了数组，方便他进行翻转，进位之后，再进行构建链表


  复杂度分析：
    时间复杂度是:O(N)
    一层循环将该链表遍历完毕。N取决与两条链表的最大长度。
    空间复杂度：O(N）
    定义了sum数组，进行数据处理，定义了res和newRes变量，存放新的和链表。N取决与两条链表的最大长度。
 */

File: BenBird/18-445两数相加II.go
//给你两个 非空 链表来代表两个非负整数。数字最高位位于链表开始位置。它们的每个节点只存储一位数字。将这两数相加会返回一个新的链表。
//
// 你可以假设除了数字 0 之外，这两个数字都不会以零开头。
//
//
//
// 进阶：
//
// 如果输入链表不能修改该如何处理？换句话说，你不能对列表中的节点进行翻转。
//
//
//
// 示例：
//
// 输入：(7 -> 2 -> 4 -> 3) + (5 -> 6 -> 4)
//输出：7 -> 8 -> 0 -> 7
//
// Related Topics 链表
// 👍 291 👎 0
package main

type ListNode struct {
	Val int
	Next *ListNode
}

type Stack struct {
	nums []int
}

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	list_one, list_two := l1, l2
	l3 := &ListNode{0, nil}
	s_one := initStack()
	s_two := initStack()

	//将链表l1入栈
	for list_one != nil {
		s_one.push(list_one.Val)
		list_one = list_one.Next
	}

	//将链表l2入栈
	for list_two != nil {
		s_two.push(list_two.Val)
		list_two = list_two.Next
	}

	carry := 0
	curr_num := 0
	curr_stack := initStack()
	for !s_one.isEmpty() || !s_two.isEmpty() || carry != 0 {
		num_one := s_one.pop()
		num_two := s_two.pop()

		sum := num_one + num_two + carry

		//控制进位数
		if sum > 9 {
			carry = 1
		} else {
			carry = 0
		}
		curr_num = sum % 10

		//将获取求和、进位处理后的数字入栈
		curr_stack.push(curr_num)
	}

	//将求和结果出栈存入链表l3
	curr_list := l3
	for !curr_stack.isEmpty() {
		curr_list.Next = &ListNode{curr_stack.pop(), nil}
		curr_list = curr_list.Next
	}

	return l3.Next
}

/**
初始化空栈
 */
func initStack() *Stack {
	return &Stack{
		nums: []int{},
	}
}

/**
入栈
 */
func (s *Stack) push(n int) {
	s.nums = append(s.nums, n)
}

/**
出栈
 */
func (s *Stack) pop() int {
	if s.isEmpty() {
		return 0
	}

	res := s.nums[len(s.nums) - 1]
	s.nums = s.nums[:len(s.nums) - 1]

	return res
}

/**
判断栈是否为空
 */
func (s *Stack) isEmpty() bool {
	if len(s.nums) == 0 {
		return true
	}

	return false
}

//leetcode submit region end(Prohibit modification and deletion)

func main() {

}

/**
题解：两数相加II

耗时：20ms
内存：6.2M

因本题的链表是从高位到低位一次排列，所以不能直接进行遍历链表依次进行求和，
所以将两链表从第一个节点开始分别存储到两个栈s_one, s_two中，
然后再同时出栈，将出栈两元素进行求和，同时控制好进位数，
但是这样求和结果是逆序的，所以先求和结果存入栈curr_stack
然后将curr_stack中的数一次出栈，存入链表中
这样一正一反，求和结果就变成了正序

复杂度分析：
	时间复杂度：O(n)
	空间复杂度：O(n)
 */
